File: src/Python/remote/http_step.py
from __future__ import annotations
import json
import time
from typing import Callable, Dict, Any
from urllib import request, error, parse

class RemoteSpec:
    def __init__(self, *, endpoint: str, timeout_millis: int = 1000, retries: int = 0,
                 headers: Dict[str, str] | None = None, method: str = 'POST'):
        self.endpoint = endpoint
        self.timeout_millis = timeout_millis
        self.retries = retries
        self.headers = headers or {}
        self.method = method.upper()
        self.to_json: Callable[[Any], str] = lambda i: json.dumps(i)
        self.from_json: Callable[[str], Any] = lambda s: json.loads(s)

def http_step(spec: RemoteSpec):
    def _call(x):
        data_str = spec.to_json(x)
        last_err: Exception | None = None
        for attempt in range(spec.retries + 1):
            try:
                if spec.method == 'GET':
                    # Build a query string from dict/JSON when possible
                    query = ''
                    try:
                        obj = json.loads(data_str)
                        if isinstance(obj, dict):
                            query = parse.urlencode(obj, doseq=True)
                        else:
                            query = str(obj)
                    except Exception:
                        # Not JSON; treat string as already-encoded query fragment
                        query = str(data_str) if data_str else ''
                    url = spec.endpoint
                    if query:
                        url = url + ('&' if ('?' in url) else '?') + query
                    req = request.Request(url, method='GET', headers=spec.headers)
                    with request.urlopen(req, timeout=spec.timeout_millis / 1000.0) as resp:
                        body = resp.read().decode('utf-8')
                        return spec.from_json(body)
                else:
                    req = request.Request(spec.endpoint, data=data_str.encode('utf-8'), method='POST', headers={
                        'Content-Type': 'application/json', **spec.headers
                    })
                    with request.urlopen(req, timeout=spec.timeout_millis / 1000.0) as resp:
                        body = resp.read().decode('utf-8')
                        return spec.from_json(body)
            except Exception as e:
                last_err = e
                if attempt < spec.retries:
                    time.sleep(0.05 * (attempt + 1))
                else:
                    raise
        if last_err:
            raise last_err
    return _call

File: src/typescript/remote/http_step.ts
export interface RemoteSpecInit {
  endpoint: string;
  timeoutMillis?: number;
  retries?: number;
  headers?: Record<string, string>;
  method?: "GET" | "POST";
  toJson?: (inputValue: unknown) => string;
  fromJson?: (text: string) => unknown;
}

export class RemoteSpec {
  endpoint: string;
  timeoutMillis: number;
  retries: number;
  headers: Record<string, string>;
  method: "GET" | "POST";
  toJson: (inputValue: unknown) => string;
  fromJson: (text: string) => unknown;

  constructor(init: RemoteSpecInit) {
    this.endpoint = init.endpoint;
    this.timeoutMillis = init.timeoutMillis ?? 1000;
    this.retries = init.retries ?? 0;
    this.headers = { ...(init.headers ?? {}) };
    this.method = (init.method ?? "POST").toUpperCase() as "GET" | "POST";
    this.toJson = init.toJson ?? ((inputValue: unknown) => JSON.stringify(inputValue));
    this.fromJson = init.fromJson ?? ((text: string) => JSON.parse(text));
  }
}

export function httpStep(spec: RemoteSpec) {
  return async function callRemote(inputValue: unknown): Promise<unknown> {
    const bodyText = spec.toJson(inputValue);
    let lastError: unknown | undefined;

    for (let attemptIndex = 0; attemptIndex < spec.retries + 1; attemptIndex += 1) {
      try {
        if (spec.method === "GET") {
          let queryString = "";
          try {
            const objectValue = JSON.parse(bodyText);
            if (objectValue && typeof objectValue === "object") {
              const params = new URLSearchParams();
              for (const [key, value] of Object.entries(objectValue as Record<string, unknown>)) {
                if (Array.isArray(value)) {
                  for (const item of value) {
                    params.append(key, String(item));
                  }
                } else if (value != null) {
                  params.append(key, String(value));
                }
              }
              queryString = params.toString();
            } else {
              queryString = String(objectValue);
            }
          } catch {
            queryString = bodyText ? String(bodyText) : "";
          }
          let url = spec.endpoint;
          if (queryString.length > 0) {
            url = url + (url.includes("?") ? "&" : "?") + queryString;
          }
          const response = await fetch(url, {
            method: "GET",
            headers: spec.headers
          });
          const text = await response.text();
          return spec.fromJson(text);
        } else {
          const response = await fetch(spec.endpoint, {
            method: "POST",
            headers: { "Content-Type": "application/json", ...spec.headers },
            body: bodyText
          });
          const text = await response.text();
          return spec.fromJson(text);
        }
      } catch (caughtError) {
        lastError = caughtError;
        if (attemptIndex < spec.retries) {
          const delayMillis = 50 * (attemptIndex + 1);
          await new Promise<void>(resolve => setTimeout(resolve, delayMillis));
          continue;
        } else {
          throw caughtError;
        }
      }
    }

    if (lastError != null) {
      throw lastError;
    }
    return inputValue;
  };
}
